### PR TITLE
Fix fetch URL for exception.json and update license files

### PR DESCRIPTION
**Pull request by fjssilva**
The script [`license-text-extractor.ps1`](https://github.com/NuGet/NuGet.Licenses/blob/main/src/NuGet.Licenses/license-text-extractor.ps1) was wrongly fetching an old version of the file `exception.json` that only contained a small number of license exceptions.

This fixes the [NuGetGallery issue #10225](https://github.com/NuGet/NuGetGallery/issues/10225). As stated in the issue, I only require the missing text for the exception `Universal-FOSS-exception-1.0` but, given how license files are updated in this repository, all missing licenses and other changes were automatically applied to this patch.

### CODE REVIEW

File: src/NuGet.Licenses/App_Data/Licenses/389-exception.json
Comment (erdembayar) [inline src/NuGet.Licenses/App_Data/Licenses/389-exception.json:3]: @fjssilva 
Could we revert back those changes existing files unless there was actual typo or mismatch? Change in existing files makes really hard to review this PR, I just compared 1 of them only change seems like new line. I feel making unnecessary changes  many documents same time is risk.

File: src/NuGet.Licenses/App_Data/Licenses/PCRE2-exception.json
Comment (erdembayar) [inline src/NuGet.Licenses/App_Data/Licenses/PCRE2-exception.json:1]: Was the license already supported by us? Otherwise, we need to triage the case by case. Not sure what this license for? Same applies for others.

File: src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json
Comment (erdembayar) [inline src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json:1]: Why this one was deleted?

File: src/NuGet.Licenses/App_Data/Licenses/PCRE2-exception.json
Comment (fjssilva) [reply src/NuGet.Licenses/App_Data/Licenses/PCRE2-exception.json:1]: I don't think this is a matter of licenses being supported or not. This repository keeps and maintain replicas of the licences and exceptions hold in https://spdx.org/licenses/  and https://spdx.org/licenses/exceptions-index.html. The script [license-text-extractor.ps1](https://github.com/NuGet/NuGet.Licenses/blob/main/src/NuGet.Licenses/license-text-extractor.ps1) in this repository does all the cloning and updating work. It was only a matter of chance (bug) that this script was stuck on downloading the exceptions from an old commit in the SPDX GitHub repository. Once I fixed the script, all newer license exception files were downloaded and processed automatically.

Same applies to all new files added in this patch. Actually, you can replicate this fix quite easily: just fix the script [license-text-extractor.ps1](https://github.com/NuGet/NuGet.Licenses/blob/main/src/NuGet.Licenses/license-text-extractor.ps1) as I did and run it.

File: src/NuGet.Licenses/App_Data/Licenses/389-exception.json
Comment (fjssilva) [reply src/NuGet.Licenses/App_Data/Licenses/389-exception.json:3]: Mind that I didn't change these files myself. This was done automatically by the script [license-text-extractor.ps1](https://github.com/NuGet/NuGet.Licenses/blob/main/src/NuGet.Licenses/license-text-extractor.ps1) which downloads all licence files from https://spdx.org/licenses/ and re-creates local copies from them by extracting the corresponding sections from the original HTML. If there are changes, then this is because the license files were changed in the origin.

File: src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json
Comment (fjssilva) [reply src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json:1]: Same as before, the script [license-text-extractor.ps1](https://github.com/NuGet/NuGet.Licenses/blob/main/src/NuGet.Licenses/license-text-extractor.ps1) did it, not me.

Comment (fjssilva) [reply src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json:1]: But, digging into the details, the reason why it was deleted is because this license is now deprecated and the extractor script skips those https://github.com/NuGet/NuGet.Licenses/blob/main/src/NuGet.Licenses/license-text-extractor.ps1#L19.

Before my fix, the list of license exceptions was wrongly taken from https://raw.githubusercontent.com/spdx/license-list-data/45d10da0366f5fa931f60f3931fd23d5fb708de5/json/exceptions.json, where `Nokia-Qt-exception-1.1` wasn't deprecated yet, and now it is deprecated in the current/latest https://raw.githubusercontent.com/spdx/license-list-data/main/json/exceptions.json.

File: src/NuGet.Licenses/App_Data/Licenses/389-exception.json
Comment (fjssilva) [reply src/NuGet.Licenses/App_Data/Licenses/389-exception.json:3]: FYI, I only fixed a bug in the script [license-text-extractor.ps1](https://github.com/NuGet/NuGet.Licenses/blob/main/src/NuGet.Licenses/license-text-extractor.ps1) and follow the instructions [Updating license text](https://github.com/NuGet/NuGet.Licenses?tab=readme-ov-file#updating-license-text).

File: src/NuGet.Licenses/App_Data/Licenses/PCRE2-exception.json
Comment (erdembayar) [reply src/NuGet.Licenses/App_Data/Licenses/PCRE2-exception.json:1]: Let's scope to `Universal-FOSS-exception-1.0 license`. We need approval for new licenses on a case-by-case basis. Let's solve one problem at the time, there're not simple documents, actual legal document, we need to be careful.

File: src/NuGet.Licenses/App_Data/Licenses/389-exception.json
Comment (erdembayar) [reply src/NuGet.Licenses/App_Data/Licenses/389-exception.json:3]: Please revert this change and create separate PR if necessary.

File: src/NuGet.Licenses/license-text-extractor.ps1
Comment (erdembayar) [inline src/NuGet.Licenses/license-text-extractor.ps1:18]: I think we should use specific latest commit hash instead of  just main.

Comment (erdembayar) [reply src/NuGet.Licenses/license-text-extractor.ps1:18]: ^ @NuGet/gallery-team
What do you think?

File: src/NuGet.Licenses/App_Data/Licenses/PCRE2-exception.json
Comment (erdembayar) [reply src/NuGet.Licenses/App_Data/Licenses/PCRE2-exception.json:1]: Some license supported by https://spdx.org/licenses/ doesn't mean we should immediately accept it.

File: src/NuGet.Licenses/license-text-extractor.ps1
Comment (fjssilva) [reply src/NuGet.Licenses/license-text-extractor.ps1:18]: Note that this change was made in line with https://github.com/NuGet/NuGet.Licenses/blob/main/src/NuGet.Licenses/license-text-extractor.ps1#L10, which also downloads the licenses list from main.

Comment (erdembayar) [reply src/NuGet.Licenses/license-text-extractor.ps1:18]: I'll wait for @joelverhagen and @agr's feedback on this since they created the extractor tool.

File: src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json
Comment (erdembayar) [reply src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json:1]: Considering we assume that packages never mutate once published, and I believe it extends to embedded data like license too. What happens if the package `Nokia123` was published with the `Nokia-Qt-exception-1.1.json` license 5 years ago and the license is deleted now? Suddenly, it turns into a package without any license. Now, any enterprise consuming that package is in limbo or has to constantly check if it still has a license. Instead, regardless of what happens, the package should always keep its original license. Please correct me if I'm wrong.

File: src/NuGet.Licenses/license-text-extractor.ps1
Comment (agr) [reply src/NuGet.Licenses/license-text-extractor.ps1:18]: The comment starting at line 14 explains why specific version has to be used for exceptions. We can't just grab latest from `main`, we have to align with the snapshot used by the version of the `NuGet.Packaging` package referenced by the project.

**Review by agr — CHANGES_REQUESTED** [review]
We can't just blindly grab the latest version of exception data. It has to align with the NuGet.Packaging supported list.

File: src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json
Comment (fjssilva) [reply src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json:1]: I don't disagree with you. But then, IMHO, the license extractor tool should be fixed to not exclude deprecated licenses. Not my call.

**Comment by fjssilva** [comment]
> We can't just blindly grab the latest version of exception data. It has to align with the NuGet.Packaging supported list. Specifically, the concern is, if I remember correctly, that we shouldn't have any exceptions client doesn't know about. And it seems the package haven't updated its list of licenses/exceptions for a while (the snapshot referenced in the code haven't changed for a while). I'll ask the client folks if they can sync the list, then we can pull the changes here and update the list on our end.
> 

So, exceptions are the issue? No problem at all in having (main) licenses the client doesn't know about? BTW, all fine by me. I'm just pointing it out because it sounds strange to me, especially because NuGetLicenseData.cs does refer to a specific license list version...

> However, it seems the package "knows" about the `Universal-FOSS-exception-1.0` exception, so maybe, just changing the URL to use the `bb0099c` snapshot, the client expects would be enough.

Yes, it seems so. However, mind that running the extractor tool with this snapshot will still remove the exception `Nokia-Qt-exception-1.1`, if I'm not mistaken.

File: src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json
Comment (erdembayar) [reply src/NuGet.Licenses/App_Data/Licenses/Nokia-Qt-exception-1.1.json:1]: We can't focus on extractor tool right now due to high priority work we have scheduled right now. This isn't ideal and you're welcome to contribute fixing the tool.

**Comment by agr** [comment]
I don't remember the details, but the API of the licenses service assumes that SPDX license expression is passed as the query path. If the expression contains more than one element, a page is displayed with links to individual elements (licenses or exceptions). However, standalone exception name is not a valid SPDX license expression (exceptions can only follow the `WITH` keyword). To work around that, the first thing the code does is comparing the "expression" with a known exception list. That known exception list comes from `NuGet.Packaging`. If that lookup fails it tries to parse the whole path as license expression, using, again NuGet.Packaging. In either case, the presence of exceptions NuGet.Packaging doesn't know about would cause a failure.

The `38193d69ba3e0b9e84f94ebc23c6e3a2bd9b0bec` snapshot from SPDX repo seems to be the right one to use in this case to align with NuGet.Packaging version the project uses.

I also looked through all the exceptions specified in license expressions for packages, and it seems none use `Nokia-Qt-exception-1.1`, so it shouldn't break anything.

**Comment by fjssilva** [comment]
Thanks @agr. FYI, this PR is a consequence of the missing page in Licensing information for the [MySQL.Data package](https://www.nuget.org/packages/MySql.Data). As you can see, the links are rendered properly but the [Universal-FOSS-exception-1.0](https://licenses.nuget.org/Universal-FOSS-exception-1.0) links to nowhere.

So, IIUC, what I should do next is to rework this patch so that it downloads the exceptions from the snapshot `38193d69ba3e0b9e84f94ebc23c6e3a2bd9b0bec` (license list version `bb0099c`) and run the license extractor tool, right?

Just one remark: if I do so, the patch will also include all the new new licenses (not only exceptions) and changes in some of the existing license files, same as what you see already on this PR since the licenses are downloaded from `main`. Is this what you expect?

**Comment by agr** [comment]
Yes, please update the exception snapshot. Pulling new licenses is fine.

For completeness, I sent the PR to update NuGet.Packaging to sync the license list to its today's state, so when the package update is published, we'd enable those to be actually used for new packages. This would require re-sync of exceptions, but, we'll sort it out later.

**Comment by fjssilva** [comment]
PR updated. The extractor tool now downloads from the snapshot `38193d69ba3e0b9e84f94ebc23c6e3a2bd9b0bec` and the tool was re-run.

**Review by agr — APPROVED** [review]
Thank you!

I verified that those are all the changes that are available in SPDX repo and spot checked several licenses that the changes brought in are indeed the license template maintenance.

**Comment by agr** [comment]
@fjssilva, you'd need to [agree](https://github.com/NuGet/NuGet.Licenses/pull/54#issuecomment-2438160102) to CLA in order for this PR to be unblocked for merging.

**Comment by erdembayar** [comment]
> @fjssilva, you'd need to [agree](https://github.com/NuGet/NuGet.Licenses/pull/54#issuecomment-2438160102) to CLA in order for this PR to be unblocked for merging.

@fjssilva 
Please check above, you need to agree CLA in order to continue.
image

**Comment by fjssilva** [comment]
@erdembayar I know that, but I can't sign the CLA without my company's (Oracle) acceptance, and for that I need to involve our legal team, which may lead to a long bureaucratic process.

**Comment by erdembayar** [comment]
> @erdembayar I know that, but I can't sign the CLA without my company's (Oracle) acceptance, and for that I need to involve our legal team, which may lead to a long bureaucratic process.

I don't know what the best way is for you. Is there anything owned by Oracle in this PR?

**Comment by fjssilva** [comment]
> @fjssilva please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

**Comment by fjssilva** [comment]
@dotnet-policy-service agree company="Oracle America, Inc."

**Comment by erdembayar** [comment]
@fjssilva 
Thank you for contribution, we'll evaluate impact on `NuGet Client` and `NuGet Gallery`, we need to make other changes to flow from here.

**Comment by agr** [comment]
This PR brings licenses service up to date with client library versions used both by the licenses service and Gallery. Until [recent PR](https://github.com/NuGet/NuGet.Client/pull/6137/files) client was using SPDX list version `bb0099c` which this PR updates to.

**Pull request merged**
